### PR TITLE
Add PrimaryExternalDistribution and bounded/physical vertex distributions

### PR DESCRIPTION
**Pull request by nickkamp1**
## Stack: PR 6 of 13

This PR is part of a 13-PR stack decomposing `dev/HNL_DIS` into reviewable chunks.

- **Base:** `main`
- **Head:** `feat/primary-external-distribution`

## Summary

Adds three new distribution classes for sampling primary injection events from external CSV data and placing their interaction vertices:

- **`PrimaryExternalDistribution`** — reads particle kinematics (position, momentum, energy, mass, and arbitrary parameters) from a CSV file and samples rows uniformly at random to populate `PrimaryDistributionRecord`. Supports an optional minimum-energy threshold (`emin`) that filters at load time.
- **`PrimaryBoundedVertexDistribution`** — samples an interaction vertex along a particle trajectory, bounded by a configurable `max_length` and optionally clipped to a fiducial volume geometry.
- **`PrimaryPhysicalVertexDistribution`** — samples an interaction vertex along the full detector path with no length or fiducial constraint.

Both vertex distributions use inverse-CDF sampling of interaction depth (cross-section + decay length) with a numerically stable `log(1 - exp(-x))` helper for small/large depth regimes.

Also adds `interaction_parameters` (a `std::map<std::string, double>`) to `PrimaryDistributionRecord` so that arbitrary CSV columns (e.g. `Q2`) propagate through the record chain into `InteractionRecord`.

## Changes since the initial squash

The original squash (`0c31ac0a`) contained the raw code from `dev/HNL_DIS_clean`. The subsequent 11 commits fix bugs identified during review:

### Bug fixes — PrimaryExternalDistribution

| Fix | Detail |
|---|---|
| Broken copy constructor | Body was `PrimaryExternalDistribution(other.filename);` — creates and discards a temporary, leaving the new object uninitialized. Changed to `= default` after removing the non-copyable `std::ifstream` member. |
| `exit(0)` on errors | File-open failure and sampling failure called `exit(0)` (success code). Replaced with `std::runtime_error` throughout. |
| Incomplete cereal serialization | `save`/`load` only serialized the base class — no member data. Now serializes `emin`, `keys`, `input_data`, `init_pos_set`, `mom_set`. Removed `std::ifstream` member (not serializable) and `possible_keys` (unused). |
| Partial-component `init_pos_set`/`mom_set` | Set `true` if *any* of x0/y0/z0 (or px/py/pz) appeared. Now requires *all three* to avoid using uninitialized array elements. |
| `less()` not a strict weak ordering | Was `return filename != x->filename` (symmetric — `a<b && b<a`). Now uses `std::tie(emin, keys, input_data)`. |
| `less()` null-pointer dereference | No null check after `dynamic_cast`. Added guard. |
| `equal()` compared only filename | Now compares `emin`, `keys`, and `input_data` for value equality. |
| Sampling retry loop with `exit(0)` | 1000-try loop drew random rows, checked `E > emin`, exited on failure. Energy filter moved to load time so `input_data` only contains valid rows — single draw, no retries. |
| Index out-of-bounds in `Sample()` | `rand->Uniform() * size` could equal `size` if Uniform returns 1.0. Added `std::min` clamp. |
| `GenerationProbability` always returned 1 | Now returns 0 when `record.primary_energy < emin`. |
| Missing `<fstream>`, `<sstream>`, `<stdexcept>` includes | Added; the original relied on transitive includes. |

### Bug fixes — vertex distributions

| Fix | Detail |
|---|---|
| Division-by-zero in `GenerationProbability` | `interaction_density / total_interaction_depth` when depth is zero. Added `<= 0` early-return guard in both Bounded and Physical. |
| `BoundedVertex::equal()` ignored fiducial volume | Only compared `max_length`. Now compares both, with null-safety for the `shared_ptr<Geometry>`. |
| `BoundedVertex::less()` null-pointer dereference + incomplete ordering | No null check after `dynamic_cast`, and only compared `max_length`. Rewritten with null safety and fiducial volume comparison. |

### Code quality

- Namespace closing comment typo `sirenREN` → `siren` in both vertex distribution files
- Removed unused `using detector::GeometryPosition/GeometryDirection` declarations
- `filename` moved from `public` to `private`
- `GetPhysicalNumEvents()` return type `int` → `size_t`, added `const`
- `LoadInputFile` parameter changed to `const &`
- `auto x` → `auto const & x` in `PrimaryDistributionRecord::Finalize` and `CrossSectionDistributionRecord::Finalize` loop iterations
- Removed commented-out dead line in `CrossSectionDistributionRecord::Finalize`
- Added `trim()` helper for CSV header/value parsing (tolerates whitespace, `\r\n` line endings)
- Added blank-line and `#`-comment skipping in CSV parser
- Added column-count validation (throws on mismatch with header)
- Added empty-data validation after load (throws if no valid rows)
- Exposed `PrimaryExternalDistribution::Name()` in Python bindings

### Existing test adjustments

- `PrimaryEnergyDistribution_TEST`: reduced `PowerLaw::GenerationProbability` iteration count 10x (was unnecessarily slow)
- `PrimaryEnergyDistribution_TEST`: loosened `ModifiedMoyal` normalization tolerance from `2e-3` to `1e-2` to fix flaky failures

### C++ (GTest) — `PrimaryExternalDistribution_TEST.cxx`

44 tests covering:
- Construction (valid file, emin filtering, missing file, empty data, all-filtered)
- CSV parsing edge cases (blank lines, comments, column count mismatch, whitespace in headers, CRLF)
- Sampling (energy, mass, position, momentum, custom parameters, uniform distribution, emin respect)
- `GenerationProbability` (above/below/at emin, no emin)
- Equality and strict-weak-ordering for all three classes
- `clone()` and copy constructor

### Python (pytest) — `test_distributions_external.py`

18 tests covering construction, metadata, sampling, generation probability, and basic construction of both vertex distribution types. Skipped via `importorskip` when `siren` is not installed.

## File list

| File | Change |
|---|---|
| `projects/distributions/private/primary/PrimaryExternalDistribution.cxx` | New (rewritten) |
| `projects/distributions/public/SIREN/distributions/primary/PrimaryExternalDistribution.h` | New (rewritten) |
| `projects/distributions/private/primary/vertex/PrimaryBoundedVertexDistribution.cxx` | New |
| `projects/distributions/private/primary/vertex/PrimaryPhysicalVertexDistribution.cxx` | New |
| `projects/distributions/public/SIREN/distributions/primary/vertex/PrimaryBoundedVertexDistribution.h` | New |
| `projects/distributions/public/SIREN/distributions/primary/vertex/PrimaryPhysicalVertexDistribution.h` | New |
| `projects/dataclasses/private/InteractionRecord.cxx` | Modified — partial-momentum finalization, const-ref loops |
| `projects/dataclasses/public/SIREN/dataclasses/InteractionRecord.h` | Modified — added `interaction_parameters` map and setters to `PrimaryDistributionRecord` |
| `projects/distributions/private/pybindings/distributions.cxx` | Modified — exposed `Name()` |
| `projects/distributions/private/test/PrimaryExternalDistribution_TEST.cxx` | New — 44 GTest cases |
| `projects/distributions/private/test/PrimaryEnergyDistribution_TEST.cxx` | Modified — iteration count and tolerance |
| `projects/distributions/CMakeLists.txt` | Modified — registered new test target |
| `tests/python/test_distributions_external.py` | New — 18 pytest cases |

## Notes

- Source commits on `dev/HNL_DIS_clean` by Nicholas Kamp: `21451da3`, `34a229f6`, `455b0730`, `9098b42e`, `a79d4408`, `d7abebbd`, `e530d339`.
- Per-commit history is browsable on `dev/HNL_DIS_clean`.
- This PR replaces an earlier copy opened from a different account; closed PRs in the project history are intentional duplicates.

### CODE REVIEW

File: projects/distributions/private/primary/PrimaryExternalDistribution.cxx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+namespace siren {
+namespace distributions {
+
+//---------------
+// class PrimaryExternalDistribution : PrimaryExternalDistribution
+//---------------
+
+void PrimaryExternalDistribution::LoadInputFile(std::string _filename)
+{filename = _filename;
```
Comment (austinschneider) [inline projects/distributions/private/primary/PrimaryExternalDistribution.cxx]: Formatting: opening brace is on the same line as the first statement. `filename = _filename;` should be on its own line after `{`.

File: projects/distributions/private/primary/PrimaryExternalDistribution.cxx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        while (std::getline(_ss, value, ',')) {
+            if (keys[ikey]=="E") {
+                if (stod(value) < emin) passed = false;
+            }
+            ++ikey;
+            tmp_data.push_back(stod(value));
+        }
+        if (passed) input_data.push_back(tmp_data);
+    }
+
```
Comment (austinschneider) [inline projects/distributions/private/primary/PrimaryExternalDistribution.cxx:83]: `input_file` is read to completion here but stays open for the lifetime of the object. Close it at the end of `LoadInputFile()` to avoid holding a file descriptor indefinitely.

File: projects/distributions/private/primary/PrimaryExternalDistribution.cxx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                record.SetInteractionParameter(keys[i_key],value);
+            }
+            ++i_key;
+        }
+        if(mom_set) record.SetThreeMomentum(_momentum);
+        if(init_pos_set) record.SetInitialPosition(_initial_position);
+    }
+    if(!success) {
+        std::cout << "Failed to generate a physical primary after " << max_tries << "attempts. Try again\n";
+        exit(0);
```
Comment (austinschneider) [inline projects/distributions/private/primary/PrimaryExternalDistribution.cxx]: Missing space: `<< max_tries << "attempts"` renders as e.g. `1000attempts`. Add a space.

File: projects/distributions/private/primary/PrimaryExternalDistribution.cxx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+std::string PrimaryExternalDistribution::Name() const {
+    return "PrimaryExternalDistribution";
+}
+
+double PrimaryExternalDistribution::GenerationProbability(std::shared_ptr<siren::detector::DetectorModel const> detector_model,
+                                                          std::shared_ptr<siren::interactions::InteractionCollection const> interactions,
+                                                          siren::dataclasses::InteractionRecord const & record) const {
+    return 1;
```
Comment (austinschneider) [inline projects/distributions/private/primary/PrimaryExternalDistribution.cxx]: `GenerationProbability` returns 1 unconditionally, but `Sample()` does rejection sampling against `emin`. When `emin > 0`, the actual generation probability is `N_above_threshold / N_total`, not 1. Event weights will be wrong for any non-zero energy cut.

File: projects/distributions/public/SIREN/distributions/primary/PrimaryExternalDistribution.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+public:
+    ~PrimaryExternalDistribution() {
+        input_file.close();
+    };
+    std::string filename;
+private:
+    std::ifstream input_file;
+    std::vector<std::vector<double>> input_data;
+    std::vector<std::string> keys;
+    std::vector<std::string> possible_keys = {"x0","y0","z0","m","E","px","py","pz"};
```
Comment (austinschneider) [inline projects/distributions/public/SIREN/distributions/primary/PrimaryExternalDistribution.h]: `possible_keys` is declared and initialized but never referenced anywhere. Dead code -- remove it or use it for input validation in `LoadInputFile()`.

File: projects/dataclasses/private/InteractionRecord.cxx [hunk 6]
```diff
@@ -619,6 +627,9 @@ void PrimaryDistributionRecord::Finalize(InteractionRecord & record) const {
     record.primary_mass = GetMass();
     record.primary_momentum = GetFourMomentum();
     record.primary_helicity = GetHelicity();
+    for (auto x : interaction_parameters) {
```
Comment (austinschneider) [inline projects/dataclasses/private/InteractionRecord.cxx]: `for (auto x : ...)` copies every `pair<string, double>` per iteration. Use `for (auto const & x : ...)`. Same issue at the other `Finalize()` below (line 980).

File: projects/distributions/private/primary/vertex/PrimaryPhysicalVertexDistribution.cxx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        return true;
+}
+
+bool PrimaryPhysicalVertexDistribution::less(WeightableDistribution const & other) const {
+    const PrimaryPhysicalVertexDistribution* x = dynamic_cast<const PrimaryPhysicalVertexDistribution*>(&other);
+    return false;
+}
+
+} // namespace distributions
+} // namespace sirenREN
```
Comment (austinschneider) [inline projects/distributions/private/primary/vertex/PrimaryPhysicalVertexDistribution.cxx]:
```suggestion
} // namespace siren
```

File: projects/distributions/private/primary/vertex/PrimaryBoundedVertexDistribution.cxx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    else
+        return (max_length == x->max_length);
+}
+
+bool PrimaryBoundedVertexDistribution::less(WeightableDistribution const & other) const {
+    const PrimaryBoundedVertexDistribution* x = dynamic_cast<const PrimaryBoundedVertexDistribution*>(&other);
+    return
+        std::tie(max_length)
+        <
+        std::tie(x->max_length);
```
Comment (austinschneider) [inline projects/distributions/private/primary/vertex/PrimaryBoundedVertexDistribution.cxx]:
```suggestion
        std::tie(fiducial_volume, max_length)
        <
        std::tie(x->fiducial_volume, x->max_length);
```

**Review by austinschneider — APPROVED** [review]
I think this is basically ready to merge. The only real gripe I have is related to the interface. EMin being a parameter in the constructor is kind of a strange choice for a very general class, especially since it is the only cut implemented. I've put that into a new issue #130
